File: basic_dsa_fifty_five.cpp
/**
 * Basic DSA - Insert into a BST (LeetCode 701)
 * Insert value into BST and return root.
 */
#include <iostream>

struct TreeNode {
    int val;
    TreeNode* left;
    TreeNode* right;
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
};

TreeNode* insertIntoBST(TreeNode* root, int val) {
    if (!root) return new TreeNode(val);
    if (val < root->val) root->left = insertIntoBST(root->left, val);
    else root->right = insertIntoBST(root->right, val);
    return root;
}

void inorder(TreeNode* n) {
    if (!n) return;
    inorder(n->left);
    std::cout << n->val << " ";
    inorder(n->right);
}

int main() {
    std::cout << std::string(60, '=') << "\n";
    std::cout << "Problem: Insert into a Binary Search Tree\n";
    std::cout << std::string(60, '=') << "\n";

    TreeNode* root = new TreeNode(4);
    root->left = new TreeNode(2);
    root->right = new TreeNode(7);
    root->left->left = new TreeNode(1);
    root->left->right = new TreeNode(3);
    insertIntoBST(root, 5);
    std::cout << "[4,2,7,1,3], val=5 -> inorder: ";
    inorder(root);
    std::cout << "\n";

    TreeNode* r2 = insertIntoBST(nullptr, 5);
    std::cout << "[], val=5 -> ";
    inorder(r2);
    std::cout << "\n";
}

File: basic_dsa_fifty_five.py
"""
Basic DSA - Search Insert Position
Given a sorted array of distinct integers and a target, return the index if target is found.
If not, return the index where it would be inserted in order.
"""


def search_insert(nums, target):
    """
    Binary search for lower_bound (first index i with nums[i] >= target).
    Time Complexity: O(log n)
    Space Complexity: O(1)
    """
    left, right = 0, len(nums)
    while left < right:
        mid = left + (right - left) // 2
        if nums[mid] < target:
            left = mid + 1
        else:
            right = mid
    return left


if __name__ == "__main__":
    print("=" * 60)
    print("Problem: Search Insert Position")
    print("=" * 60)

    tests = [
        ([1, 3, 5, 6], 5, 2),
        ([1, 3, 5, 6], 2, 1),
        ([1, 3, 5, 6], 7, 4),
        ([1, 3, 5, 6], 0, 0),
        ([1], 0, 0),
        ([1], 1, 0),
    ]

    for nums, target, expected in tests:
        got = search_insert(nums, target)
        print(f"nums={nums}, target={target} -> {got} (expected {expected})")
